OpenKruise Project Security Self-Assessment - Security Pals

bradcush opened the pull request:
Created and added first draft for OpenKruise Project Security Self-Assessment.
Please feel free to share your feedback on the security self-assessment.

bradcush commented:
@eddie-knight Thanks for starting the review and for the initial comments!

I've included a manually generated SBOM with the assessment for reference at the time the security assessment was completed because the project itself does not generate on during the build or [release process](https://github.com/openkruise/kruise/releases). They only have `go.mod` files that are associated with the dependencies of the project. Do you still think we should link nothing there in the meantime until the project has one?

ragashreeshekar reviewed (changes requested):
Thanks for the good work @bradcush and team, appreciate the efforts. 
I have completed first pass of review and there are two things that need your attention. Please feel free to reach out here or on slack for any questions and clarifications. Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

ragashreeshekar commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,656 @@
+free Express Learning course provided by Linux Foundation Training &
+Certification: [Security Assessments for Open Source Projects](https://training.linuxfoundation.org/express-learning/security-self-assessments-for-open-source-projects-lfel1005/).
+
+# Self-assessment outline
```
All web content should only have a single H1 entry (#)

bradcush replied:
I've updated the markdown file to only have one H1 entry

bradcush commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,656 @@
+
+##### Recommended Mitigations
+
+- Assign proper roles (per principle of least privilege) to the users using
```
Thanks for raising this point. Let me take a closer look at get back to you here.

bradcush replied:
So to clarify bit here, the threat model is OpenKruise specific. When it comes to some of the more general points in the threat model like the example you listed around RBAC in Kubernetes, we would assume that they are following the principle of least privilege here but we we're able to prove that it was in fact the case. For that reason we've made sure to explicitly list it as a point of concern here since it's not clear for the project. The same would go for a few of the other more general points.

We have had one of the maintainers take a look at the entire threat model so my assumption is also that if this wasn't flagged as irrelevant then it makes sense to keep. What are your thoughts on how to proceed there?

ragashreeshekar replied:
Got it. Glad to hear maintainers have reviewed this work and are good with it. Sounds good to me.
Following discussions among the review team, we are asking that all STRIDE content be moved to a secondary document within this directory. Please extract this content and add a link to the new file in the appendix.

bradcush commented:
> Thanks for the good work @bradcush and team, appreciate the efforts. I have completed first pass of review and there are two things that need your attention. Please feel free to reach out here or on slack for any questions and clarifications. Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

Thanks @ragashreeshekar for a timely first review. I'll let you know as soon as these issues are fixed and we're ready for another review. I'll also make sure to get some clarification on some of the points raised regarding the STRIDE thread modeling.

bradcush commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,656 @@
+# OpenKruise Self-assessment
+
+The Self-assessment is the initial document for projects to begin thinking
```
I've gone through the document and removed all template text. There is one bit of text in the document under the "Self-assessment use" section which I've decided to keep as it gives a bit more context on the document itself. It will be common across all the self-assessments but I felt this context still made sense to keep.

bradcush commented:
@ragashreeshekar I think we're ready for another look when you have time. I've addressed two out of your three comments and rebased the branch on top of the latest `main`. We still have the open question about [how to handle SBOMs](https://github.com/cncf/tag-security/pull/1174#issuecomment-1847339705) which the project doesn't generate currently. (This is actually recommended as part of our assessment for that reason) I'll defer to you on what we should do next for our threat modeling also but I've responded to your concern.

eddie-knight commented:
Hey @bradcush, thanks for the explanation! 

The reviewer team had a chat going regarding the SBOMs, since this was a recurring thing we saw from other Security Pals contributions.

We should absolutely be highlighting known gaps in the self assessment, such as this, by stating that it is not currently addressed. This will help projects know gaps that they should address before continuing on to a joint assessment.

bradcush commented:
> We should absolutely be highlighting known gaps in the self assessment, such as this, by stating that it is not currently addressed. This will help projects know gaps that they should address before continuing on to a joint assessment.

Ok great, thanks for looking into it @eddie-knight. As we have it listed the creation of an SBOM as a recommendation already in the self-assessment, should I go ahead and remove the local one I have generated or should we keep that until the project officially generates one as part of their build/release? Sorry for the back and forth, just want to be clear here.

eddie-knight commented:
> just want to be clear here

Always appreciate back-and-forth for clarity. Please pull out the SBOM that you created, and adjust the SBOM metadata value to `OpenKruise does not currently generate an SBOM on release`.

Raga and I are still going through the rest of our review checklist for all of the recent contributions, but this PR seems to be very close to being ready for merge. We'll let you know if there is anything else that needs addressed. Thanks for continuing to work with us on this!

bradcush commented:
> Always appreciate back-and-forth for clarity. Please pull out the SBOM that you created, and adjust the SBOM metadata value to `OpenKruise does not currently generate an SBOM on release`.

Ok great, thanks! I've removed the SBOM and added the above line.

eddie-knight commented on assessments/projects/openkruise/self-assessment.md:
```suggestion
```

Following discussions among the review team, we are asking that Security Pals please remove any recommendations from the scope of this document. You may optionally communicate recommendations for projects in a more accessible location, such as in a project community channel or GitHub Issue.

bradcush replied:
Ok sounds good, we've communicated recommendations in [this issue in the OpenKruise repository](https://github.com/openkruise/kruise/issues/1463). I've gone ahead and removed all recommendations, including those that were inline for some of the sections and each those that were part of each section of the STRIDE thread modeling. The removal is in a separate commit. I'll leave resolving this conversation to you if you are happy with the changes.

eddie-knight commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+
+###### Actors
+
+- `kruise-operator` - The person administrating or operating the system
```
Is this a term used by the Kruise team? I'm quite surprised if that's the case, because [Kubernetes Operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) are already a well established thing. Instead of saying "operator" to describe the person using the system, consider simply saying "user."

bradcush replied:
This term was originally introduced by us. I've changed `krusie-operator` to user where we've needed to reference it in the document to make sure there is no confusion.

eddie-knight commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+
+###### Workflow
+
+1.  `kruise-operator` executes `kubctl-kruise` to initiate manual action
+    1.  Examples of manual action are scale and rollout
+2.  `kubctl-kruise` calls the `kruise-manager` which is part of the Control Plane
+3.  `kruise-daemon` calls the `kruise-manager` through the `kube-apiserver`
+4.  `kruise-daemon` executes an operation on the node
+    1.  Example include pre-image download and container restarting
```
This is the action. The previous two sections (Operator based action & workflow) are not serving to increase clarity of the explanation, and can be omitted. Giving the action a title is fine, but the extra text only adds additional reading.

Also consider using plain english, to reduce the time it takes for readers to parse the bullets here.

Example (possibly inaccurate):

> When a user executes a `kubectl-kruise` command, such as `scale` or `rollout`, the CLI will call on _kruise-manager_, which is part of the control plane. Simultaneously, _kruise-daemon_ will call the _kruise manager_ through the _kube-apiserver_. Finally, _kruise-daemon_ will execute an operation on the node, such as to download an image or restart a container.

eddie-knight replied:
The same suggestion also applies to the content below. Remove the fluff text and consider simplifying the bullets into clear action statements.

bradcush replied:
I've gone ahead and only kept the two main actions which were described properly. They have been kept as "Manual command" and "InPlace Update". I've also removed the additional sections you mentioned which didn't help to increase the clarity of this section. This included the removal of the small section restating some of the actors already mentioned in the previous section.

Actions have also been rewritten from a bulleted-list to use more plain English with clear action statements. The example you provided above was indeed mostly accurate so I've used it with just a few minor changes when reworking the first action.

eddie-knight commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+
+### Self-assessment use
+
+This self-assessment is created by the OpenKruise team to perform an internal
```
Was this created by the OpenKruise maintainer team? Adjust accordingly if not. Consider listing the authors and contributors in the first line of the document, immediately following the title.

bradcush replied:
Thanks for catching, I've updated this section slightly. It now more accurately states at the beginning:
> This self-assessment is created by group of Security Pals with help from the maintainers of OpenKruise to perform an analysis of the project's security.

bradcush commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+
+The operator is the person administrating or operating the system. This person
+has authorization to access the OpenKruise system and perform actions. The
+operator interacts with OpenKruise using the CLI or client SDK.
```
Makes sense, I've removed mention of `kruise-operator` and any sub-actors under that bullet point, `kubelet`, and "pods and containers" from the actors section as they are all external.

bradcush commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+##### `kubelet`
+
+The `kubelet` runs on each node. It's the primary "node-agent" that can register
+the node with the apiserver so that pods on that node can be managed.
```
Mentioned in the comment above I've removed `kubelet` from the actors. To your point it's still mentioned to properly describe the action that it is a part of.

bradcush commented on assessments/projects/openkruise/self-assessment.md:
```diff
@@ -0,0 +1,563 @@
+validates that the request corresponds to a file the client is authorized to
+access, and then returns a token to the client. The client then transmits that
+token to the file server, which, after confirming its validity, returns the file.
+
```
Thanks for catching, I've updated this section slightly. It now more accurately states at the beginning:
> This self-assessment is created by group of Security Pals with help from the maintainers of OpenKruise to perform an analysis of the project's security.

bradcush commented:
@eddie-knight Thanks again for the thorough review. I've gone ahead and addressed all of the comments and pushed the changes in another separate commit so they are easily visible. I can squash commits again before it's ready to merge if needed. I've still kept all conversations unresolved. Feel free to give it another look when you have time to make sure you're aligned with the changes I've made. Thanks!

bradcush commented:
> I can squash commits again before it's ready to merge if needed. I've still kept all conversations unresolved.

I've also just squashed the commits that were making changes based on the feedback that's been provided during to review to make sure everything is as ready as possible to merge when it's time.

ragashreeshekar commented:
Thank you for the updates. Looks good to me.
Tagging @eddie-knight for your review, and @JustinCappos both for your opinion on including maintainers for their ack and review of this work, followed by chair review (as applicable) and merging them to the repo.

JustinCappos commented:
Okay, as for merging, let's talk about a process in the upcoming TL and TAG
Security meetings.  I think having a small group that goes and completes
this process would be the best way to move forward.  We should be able to
get most of these merged with a few hours of effort, I think.

On Tue, Jan 2, 2024 at 5:31 PM Raga ***@***.***> wrote:

> Thank you for the updates. Looks good to me.
> Tagging @eddie-knight <https://github.com/eddie-knight> for your review,
> and @JustinCappos <https://github.com/JustinCappos> both for your opinion
> on including maintainers for their ack and review of this work, followed by
> chair review (as applicable) and merging them to the repo.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cncf/tag-security/pull/1174#issuecomment-1874644273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGROD4VIKB4IQDDP7GQRADYMSDEZAVCNFSM6AAAAABAJYIRT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNZUGY2DIMRXGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bradcush commented:
> Thank you for the updates. Largely looks good to me. Just one from my end on extracting threat model to a separate file within the same folder.

Thanks for taking a look. I've extracted the STRIDE Threat Model into a a new file named `threat-model.md` and linked to that file from the same section of the self assessment where it was previously inline.

eddie-knight commented:
LGTM

JustinCappos reviewed (approved):
This looks good to me.  Thanks to all for the effort!
